README.md
# README


[![Build Status](https://travis-ci.org/turion/sonnendemo.svg?branch=master)](https://travis-ci.org/turion/sonnendemo)

This is a simple example game for the gloss bindings
for [Rhine](https://github.com/turion/rhine),
an asynchronous functional reactive programming library.

## Installation

I'm assuming that you have a regular Haskell development environment, on any platform.

```
git clone https://github.com/turion/sonnendemo
cd sonnendemo
```

### With cabal

```
cabal sandbox init
cabal install --only-dependencies
cabal build
```

### With stack

```
stack build
```


## Play

* __With cabal__: 
To play it, simply run `cabal run sonnendemo` for a GUI version,
or `cabal run sonnendemo-console` for a console version.

* __With stack__:
Run `stack exec sonnendemo` for a GUI version,
or `stack exec sonnendemo-console` for a console version.

### The game (`SonnenModel.hs`)

* The objective of the game is to drink coffee.
  You cannot lose, or die.
* The coffee machine runs on renewable energy, of course,
   and is powered by a little solar plant and a wind turbine.
* The energy is stored in a battery.
  When there is a sufficient charge level in it, you can make a coffee, and then drink it.
* Under usual conditions, the battery can never be fully charged or emptied.
  There is always a little reserve charge and a little extra capacity left.
  _This is on purpose._ Sometimes, the wind will be very strong, and extra energy needs to be absorbed.
  In other times, the wind is still, and energy needs to be delivered from the battery to the grid.
  These things happen automatically. You can simply make coffees and marvel while it happens.

### GUI version (`MainGloss.hs`)

1. If there is enough energy in the battery (and a little reserve),
  a green check mark will appear next to the coffee cup.
  (Otherwise, a red cross will be displayed.)
  You can then brew a coffee by clicking on the cup.
2. To drink the coffee, click on the cup once it's full.
3. Repeat :)
4. Exit by pressing `Escape`.

Here is a screenshot:

![screenshot](screenshot.png)

### Console version (`MainConsole.hs`)

All relevant information appears on the console.
Brew and drink coffee by pressing `Enter`.


ChangeLog.md
# Revision history for sonnendemo

## 0.4.0.0

* Added wind turbine.
* Reorganised and documented code better
* Added contours to graphics

## 0.3.0.0

## 0.2.0.0  -- 2017-11-29

* First version. Version numbers follow rhine.
* Added playable demo.


MainConsole.hs
{-# LANGUAGE DataKinds       #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE TypeFamilies    #-}
module Main where

-- base
import Data.Char (toLower)

-- rhine
import FRP.Rhine
import FRP.Rhine.Clock.Realtime.Stdin
import FRP.Rhine.Clock.Realtime.Millisecond
import FRP.Rhine.ResamplingBuffer.Collect
import FRP.Rhine.ResamplingBuffer.KeepLast
import FRP.Rhine.Schedule.Concurrently

-- sonnendemo
import SonnenModel


humanPrintWeather :: Weather -> String
humanPrintWeather Weather {..} = humanPrintSun sun ++ ", " ++ humanPrintWind wind
  where
    humanPrintSun  sun    = "It's " ++ (toLower <$> show sun)
    humanPrintWind Normal = "and the wind is normal."
    humanPrintWind Weak   = "but the wind is low! We need to pump energy in the grid!"
    humanPrintWind Strong = "but the wind is strong! We need to absorb energy from the grid!"

-- | The method to print a status message.
consoleOutput :: ModelState -> IO ()
consoleOutput ModelState {..} = do
  putStrLn $ humanPrintWeather weather
  putStrLn $ "The battery is charged at "
          ++ show (round $ batteryLevel * 100 / batteryCapacity)
          ++ "%."
  putStrLn $ case coffeeState of
    Empty      -> if batteryLevel < batteryBalancingMargin + coffeeEnergy
                    then "Your coffee is empty and the battery is low... :("
                    else "How about a nice brew? Just press Enter!"
    Brewing  _ -> "Brewing..."
    Full       -> "That's looking like a nice cuppa! Press Enter to enjoy!"
    Drinking _ -> "Mmmmmhhh..."

-- | The game loop.
game
  :: SyncSF IO (RescaledClockFloat (Millisecond 40)) [a] ModelState
game = arr (not . null) >-> gameLogic

-- | The console output loop.
console :: SyncSF IO (RescaledClockFloat (Millisecond 2000)) ModelState ()
console = arrMSync consoleOutput

mainRhine
  =   syncId             @@  rescaleClockFloat StdinClock
  >-- collect            -@- concurrently
  --> game               @@  rescaleClockFloat waitClock
  >-- keepLast initModel -@- concurrently
  --> console            @@  rescaleClockFloat waitClock
  where
    initModel = ModelState
      { coffeeState  = Empty
      , batteryLevel = 0
      , weather      = Weather Sunny Normal
      , nCoffees     = 0
      }

main :: IO ()
main = flow mainRhine


-- * Helpers to deal with 'Float's instead of 'Double's as time differences.

-- | A newtype for 'UTCTime' with 'Float' instead of 'Double' as time difference.
newtype UTCTimeFloat = UTCTimeFloat UTCTime

instance TimeDomain UTCTimeFloat where
  type Diff (UTCTimeFloat) = Float
  diffTime (UTCTimeFloat t1) (UTCTimeFloat t2) = realToFrac $ diffTime t1 t2

-- | A clock rescaled to the 'TimeDomain' 'UTCTimeFloat'.
type RescaledClockFloat cl = RescaledClock cl UTCTimeFloat

-- | The rescaled clock value.
rescaleClockFloat :: TimeDomainOf cl ~ UTCTime => cl -> RescaledClockFloat cl
rescaleClockFloat cl = RescaledClock
  { rescale = UTCTimeFloat
  , unscaledClock = cl
  }


SonnenModel.hs
{- | The backend-agnostic model for simulating the virtual power plant.
To verify that the model can be simulated with different clocks,
the type 'BehaviourF' is used, which is clock-polymorphic.
-}

{-# LANGUAGE Arrows          #-}
{-# LANGUAGE RankNTypes      #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE TypeFamilies    #-}
module SonnenModel where


-- rhine
import FRP.Rhine
import FRP.Rhine.SyncSF.Except

-- * General type synonyms

type Time        = Float
type Power       = Float
type Energy      = Float
type CoffeeLevel = Float

-- * Components of the simulation

-- ** Coffee cup and machine

-- | Modelling the current state of the coffee cup.
data CoffeeState
  = Empty
  | Brewing CoffeeLevel
  | Full
  | Drinking CoffeeLevel
  deriving (Eq, Show)

isEmpty :: CoffeeState -> Bool
isEmpty Empty = True
isEmpty _     = False

-- | The level to which the cup is filled with coffee.
coffeeLevel :: CoffeeState -> Float
coffeeLevel Empty        = 0
coffeeLevel (Brewing x)  = x
coffeeLevel Full         = 1
coffeeLevel (Drinking x) = 1 - x


brewingTime :: Time
brewingTime = 2

drinkingTime :: Time
drinkingTime = 2

-- | The energy needed to brew one cup of coffee.
coffeeEnergy :: Energy
coffeeEnergy = brewingTime * batteryMaxPower


-- | The different states of the coffee machine,
--   depending on user coffee requests and battery charge level.
coffeeStates
  :: (Monad m, TimeDomain td, Diff td ~ Float)
  => BehaviourFExcept m td  (Bool, Energy) CoffeeState Empty
coffeeStates = do
  -- In case there is sufficient battery, start brewing a coffee when requested
  try $ proc (coffeeRequest, batteryLevel) -> do
    _ <- throwOn () -< coffeeRequest
                    && batteryLevel >= coffeeEnergy + batteryBalancingMargin
    returnA         -< Empty
  -- Output the progress of the brewing process until the time is up
  try $ scaledTimer brewingTime >>> arr Brewing
  -- Start drinking the coffee when requested
  try $ proc (drinkRequest, _) -> do
    _ <- throwOn () -< drinkRequest
    returnA         -< Full
  -- Output the progress of the drinking process until the time is up
  try $ scaledTimer drinkingTime >>> arr Drinking
  coffeeStates

-- ** The battery

batteryCapacity :: Energy
batteryCapacity = 4

-- | The maximum power that can be drained from or charged to the battery.
batteryMaxPower :: Power
batteryMaxPower = 0.3

-- | The minimum time under which the battery is required to be able to
--   supply primary control/balancing power.
batteryBalancingTime :: Time
batteryBalancingTime = 4

-- | The minimum charge that must remain in the battery
--   in order to supply primary control/balancing power.
batteryBalancingMargin :: Energy
batteryBalancingMargin = batteryMaxPower * batteryBalancingTime


-- | Simulate the current battery charge level,
--   depending on the current solar power and the state of the coffee machine.
batterySim
  :: (Monad m, TimeDomain td, Diff td ~ Float)
  => BehaviourF m td ((Weather, CoffeeState), Energy) Energy
batterySim = proc ((Weather {..}, coffeeState), batteryLevel) -> do
  let
    coffeeDrain = case coffeeState of
      Brewing _ -> batteryMaxPower
      _         -> 0
    solarInflux = if batteryLevel >= batteryCapacity - batteryBalancingMargin
      then 0
      else solarPlant sun
    windInflux = if batteryLevel >= batteryCapacity - batteryBalancingMargin
                 && wind /= Strong
      then -batteryMaxPower
      else windTurbine wind
    batteryTotalPower = solarInflux + windInflux - coffeeDrain
  integral -< batteryTotalPower


-- ** The weather

-- | The complete current state of the weather
data Weather = Weather
  { sun  :: Sun
  , wind :: Wind
  }
  deriving Show

-- | Simulates the weather changing over time
theWeather
  :: (Monad m, TimeDomain td, Diff td ~ Float)
  => Behaviour m td Weather
theWeather = proc _ -> do
  sun  <- safely sunStates  -< ()
  wind <- safely windStates -< ()
  returnA                   -< Weather {..}


-- | The possible states the sun can be in.
data Sun = Sunny | Cloudy | Night
  deriving Show

-- | Cycles through different sun states every 12 time steps.
--   The states were randomly selected by me ;)
sunStates
  :: (Monad m, TimeDomain td, Diff td ~ Float)
  => BehaviourFExcept m td a Sun Empty
--BehaviourFExcept m Float a Weather Empty
sunStates = do
  sequence_ [ try $ timer 12 >>> arr (const weather)
            | weather <- [Sunny, Night, Cloudy, Night]
            ]
  sunStates


-- | The possible wind strengths
data Wind = Normal | Strong | Weak
  deriving (Show, Eq)


windStates
  :: (Monad m, TimeDomain td, Diff td ~ Float)
  => BehaviourFExcept m td a Wind Empty
windStates = do
  try $ timer 7.5                  >>> arr (const Normal)
  try $ timer batteryBalancingTime >>> arr (const Strong)
  try $ timer 9.5                  >>> arr (const Normal)
  try $ timer batteryBalancingTime >>> arr (const Weak)
  windStates

-- ** The power plants

-- | Determines the power of the solar plant depending on the sun.
solarPlant :: Sun -> Float
solarPlant Sunny  = batteryMaxPower * 0.5
solarPlant Cloudy = batteryMaxPower * 0.2
solarPlant Night  = 0

-- | Determines the power (or lack thereof)
--   of the wind turbine depending on the weather.
windTurbine :: Wind -> Float
windTurbine Normal = batteryMaxPower * 0.2
windTurbine Strong = batteryMaxPower * 0.5
windTurbine Weak   = batteryMaxPower * (-0.5)

-- * Putting everything together


data ModelState = ModelState
  { coffeeState  :: CoffeeState
  , batteryLevel :: Energy
  , weather      :: Weather
  , nCoffees     :: Integer
  }

-- | The logic of the whole model.
--   Requests to make coffees can be input
--   (where 'True' represents a request, and 'False' no request).
--   The current state of the coffee machine and cup,
--   the battery charge level, the current weather,
--   and the number of drunk coffees are output.
gameLogic
  :: (Monad m, TimeDomain td, Diff td ~ Float)
  => BehaviorF m td Bool ModelState
gameLogic = feedback 0 $ proc (coffeeRequest, batteryLevelOld) -> do
  weather       <- theWeather          -< ()
  coffeeState   <- safely coffeeStates -< (coffeeRequest, batteryLevelOld)
  drunkCoffee   <- edgeTo Empty Empty  -< coffeeState
  nCoffees      <- sumS                -< if drunkCoffee then 1 else 0
  batteryLevel  <- batterySim          -< ( (weather, coffeeState)
                                          , batteryLevelOld)
  returnA                              -< ( ModelState {..}, batteryLevel )


-- * Utilities, to be ported to dunai or rhine

{- |
Emits 'True' when the input value changes
from the first given argument to the second given argument.

Note: @edgeFromTo a1 a2 a2@ will not trigger an edge on the first tick,
whereas @edgeFromTo a1 a2 a1@ will trigger an edge on the first tick.
-}
edgeFromTo
  :: (Monad m, Eq a)
  => a -- ^ The old value that the signal should have before the edge
  -> a -- ^ The new value that the signal should have _now_ to trigger the edge
  -> a -- ^ The initialisation value of the delay (the signal value at tick -1).
  -> BehaviourF m td a Bool
edgeFromTo aOld aNew aInit = proc a -> do
  aPrevious <- delay aInit -< a
  returnA                  -< a == aNew && aPrevious == aOld

-- | Emits 'True' when the input value changes
--   to the given argument, from any other value.
edgeTo
  :: (Monad m, Eq a)
  => a -- ^ The new value that the signal should have _now_ to trigger the edge
  -> a -- ^ The initialisation value of the delay (the signal value at tick -1).
  -> BehaviourF m td a Bool
edgeTo aNew aInit = proc a -> do
  aPrevious <- delay aInit -< a
  returnA                  -< a == aNew && aPrevious /= aNew


MainGloss.hs
{- | This file contains the gloss backend to the game logic.
Game events (mouse clicks) are captured
-}

{-# LANGUAGE Arrows            #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE FlexibleContexts  #-}
{-# LANGUAGE RecordWildCards   #-}
{-# LANGUAGE TypeFamilies      #-}
module Main where

-- gloss
import Graphics.Gloss.Interface.Pure.Game -- TODO This should be reexported by rhine-gloss

-- dunai
import Data.VectorSpace

-- rhine
import FRP.Rhine

-- rhine-gloss
import FRP.Rhine.Gloss

-- sonnendemo
import SonnenModel
import Util


-- * Important graphical elements

backgroundColor :: Color
backgroundColor = mixColors 10 20 green white

-- ** In the house

-- *** The coffee cup

-- | The position of the coffee cup on the screen.
coffeePos :: Point
coffeePos = (160, 0)

coffeeSize :: Vector
coffeeSize = (50, 100)

-- | The size of the coffee cup rectangle that responds to mouse clicks.
coffeeCupSize :: Vector
coffeeCupSize = coffeeSize ^+^ (30, 30)

-- | Determines whether a mouse click is inside the coffee cup.
onCoffee :: Point -> Bool
onCoffee pos = abs x < fst coffeeCupSize && abs y < snd coffeeCupSize
  where
    (x, y) = pos ^+^ (0, snd coffeeSize) ^-^ coffeePos -- Relative position

-- * The different graphical elements

-- | Draw a cup of coffee, filled according to the given 'CoffeeState'.
coffeeCup :: CoffeeState -> Picture
coffeeCup coffeeState = contoured 2 $ pictures
  -- The cup
  [ translate (-40) 60 $ color white $ thickCircle 50 10
  , color white $ uncurry rectangleUpperSolid coffeeCupSize
  -- The coffee
  , color (dark $ dark $ dark $ dark yellow)
    $ translate 0 10
    $ rectangleUpperSolid (fst coffeeSize)
    $ snd coffeeSize * coffeeLevel coffeeState
  ]

-- *** The battery

batteryPos :: Point
batteryPos = (0, 0)

batterySize :: Vector
batterySize = (30, 100)

-- | Draw a battery, filled to a given energy level.
battery :: Energy -> Picture
battery batteryLevel = contoured 2 $ pictures
  [ color (dark blue) $ uncurry rectangleUpperSolid $ batterySize ^+^ (30, 30)
  , color (greyN 80) $ translate 0 10 $ uncurry rectangleUpperSolid batterySize
  , color (light blue)
    $ translate 0 10
    $ rectangleUpperSolid (fst batterySize)
    $ snd batterySize * batteryLevel / batteryCapacity
  ]

-- *** The house

-- | A house containing the battery and the coffee machine.
house :: Picture
house = contoured 2 $ pictures $
  [ scale x 1 $ color (greyN 0.3) $ pictures
    [ translate 0 (d/2) $ rotate 90 $ rectangleUpperSolid d w
    , translate w 0 $ rectangleUpperSolid d h
    ]
  | x <- [1, -1]
  ]
  ++
  [ scale x 1 $ color (dark red) $ translate 0 (h + w)
    $ rotate (-135) $ translate 0 (-d/2) $ rectangleUpperSolid d 260
  | x <- [1, -1]
  ]
  where
    h = 140
    w = 160
    d = 10


-- ** The power plants

-- *** The solar plant

solarPowerPos :: Point
solarPowerPos = (-150, 200)


-- | Draw a little solar plant, with wires.
solarPower :: Picture
solarPower = pictures
  [ rotate 180 $ wire down
  , translate 0 (wireThickness / 2 - down)
    $ rotate 270 $ wire
    $ fst solarPowerPos - fst batteryPos + fst batterySize
  , color (dark blue) $ polygon [ h ^+^ w, h ^-^ w, negate (h ^+^ w), w ^-^ h ]
  ]
  where
    h    = (0 ,  10)
    w    = (50, -30)
    down = snd solarPowerPos - snd batteryPos - snd batterySize / 2


-- *** The wind turbine

windTurbineHeight :: Float
windTurbineHeight = 100

-- | Horizontal position of the wind turbine
windTurbinePosX :: Float
windTurbinePosX = -300

-- | Draw a picture of the wind turbine where the rotors are at the given angle.
windTurbinePicture :: Float -> Picture
windTurbinePicture angle = pictures
  [ translate 0 10 $ rotate (-90) $ wire $ windTurbinePosX - fst batteryPos
  , contoured 2 $ color (greyN 0.95) $ rectangleUpperSolid 15 100
  , contoured 2 $ translate 0 100 $ rotate angle $ color white $ pictures
      $ circleSolid 15 :
      [ rotate (120 * n) $ rectangleUpperSolid 15 80
      | n <- [1,2,3]
      ]
  ]

-- | Determines how many degrees per second the wind turbine will turn.
windTurbineSpeed :: Wind -> Float
windTurbineSpeed Normal = 50
windTurbineSpeed Strong = 100
windTurbineSpeed Weak   = 5

-- *** The wiring

wireThickness :: Float
wireThickness = 5

-- | Draw a wire of a given length
wire :: Float -> Picture
wire = color (greyN 0.05) . rectangleUpperSolid wireThickness


-- *** The weather

-- | Draw a picture for the current state of the sun.
sunPicture :: Sun -> Picture
sunPicture Sunny  = color yellow $ pictures
  $ circleSolid 40 : [ rotate (45 * n) $ rectangleSolid 5 140 | n <- [0..3] ]
sunPicture Cloudy = translate (-100) 0 $ color (greyN 0.5) $ pictures
  [ translate (-40)  0 $ circleSolid 20
  , translate   35  20 $ circleSolid 10
  , translate (-15) 10 $ circleSolid 30
  , translate   15  10 $ circleSolid 30
  , translate   45   0 $ circleSolid 20
  ]
sunPicture Night  = translate 100 0 $ pictures
  [ color (light yellow) $ circleSolid 40
  , translate (-20) (-10) $ color backgroundColor $ circleSolid 40
  ]

-- ** Animated descriptions

dancingArrow :: Monad m => Behaviour m Float Picture
dancingArrow = proc _ -> do
  time <- timeInfoOf absolute -< ()
  returnA                     -< contoured 2 $ greyN 0.7 $ pictures
    [ rotate (-135) $ translate (-d/2) $ rectangleUpperSolid d h
    , rotate   225  $ translate (-d/2) $ rectangleUpperSolid d h
    , rotate   180  $ rectangleUpperSolid d l
    ]
  where
    d = 20
    h = 50
    l = 60

description :: Monad m => BehaviourF m Float String Picture
description = proc string -> do
  arrow <- dancingArrow -< ()
  returnA               -< pictures
    [ arrow
    , translate 20 0 $ scale 0.2 0.2 $ string
    ]

-- | @slowly as dt@ returns the first @n@ elements of @as@ after the time @n * dt@.
slowly :: (Monad m, RealFrac (Diff td)) => [a] -> Diff td -> Behaviour m td [a]
slowly as dt = proc _ -> do
  time <- integrate -< 1 / dt
  returnA           -< take (round time) as


-- | When playing for the first times, give helpful text descriptions.
tutorialDescriptions :: BehaviorF m Float ModelState Picture
tutorialDescriptions = safely $ do
  try $ arr (batteryLevel >>> (>= coffeeEnergy + batteryBalancingMargin))
    >>> throwOn ()
    >>> slowly "Welcome to sonnendemo!\n Wait until the battery is charged..." 0.2
    >>> description
  try $ proc ModelState { weather = Weather {..} } -> do
    time <- timeSinceSimStart -< ()
    _    <- throwOn ()        -< time > 4 && wind == Strong
    description <<< slowly "Click on the cup to make a coffee!" 0.2 -< ()
  -- ...
  safe $ arr $ const Blank

forAtLeast :: Diff td -> BehaviorFExcept m td a b e -> BehaviorF (ExceptT e m) td a b
forAtLeast dt e = runMSFExcept $ timer dt >> e

-- ** Combining everything

-- | Combine all graphics into one picture.
graphics :: Monad m => BehaviourF m Float ModelState Picture
graphics = proc ModelState { weather = Weather {..}, .. } -> do
  windTurbineAngle <- integral <<< average 0.3 -< windTurbineSpeed wind
  returnA                                      -< translate 0 (-200) $ pictures
    [ uncurry translate coffeePos $ pictures
      [ coffeeCup coffeeState
      -- Status symbol whether a coffee can be made
      , translate 0 (fst coffeeSize) $
          if batteryLevel < batteryBalancingMargin + coffeeEnergy && isEmpty coffeeState
            then color red $ pictures -- A cross (✘)
              [ rotate (40 * sign) $ rectangleSolid 10 100 | sign <- [1, -1] ]
            else translate 65 (-20) $ color green $ scale 0.8 0.8 $ pictures -- A check mark (✔)
              [ polygon [(-20,  20), (0, 0), (0, 15)]
              , polygon [( 40, 100), (0, 0), (0, 15)]
              ]
      ]
    , translate 0 400 $ sunPicture sun
    , uncurry translate solarPowerPos solarPower
    , translate windTurbinePosX 0 $ windTurbinePicture windTurbineAngle
    , uncurry translate batteryPos $ battery batteryLevel
    , translate 100 (-20) $ house
    , translate 200 440
      $ scale 0.2 0.2 $ text $ "Coffees: " ++ show nCoffees
    ]


-- * The main program

-- | The main 'SyncSF' governing events, game logic and graphics.
--   An event is produced whenever the user has clicked on the cup at least once.
game :: GlossSyncSF ()
game = arr (not . null) >>> gameLogic >>> graphics


-- | The main 'Rhine' with event selection.
glossRhine :: GlossRhine ()
glossRhine = buildGlossRhine select game
  where
    select (EventKey (MouseButton LeftButton) Down _ pos) | onCoffee pos = Just ()
    select _ = Nothing

main :: IO ()
main = flowGloss (InWindow "sonnen" (800, 600) (10, 10)) backgroundColor 30 glossRhine
